esp-hal/src/uart/uhci.rs

```diff
@@ -0,0 +1,525 @@
+    /// `Uhci` instance and the associated buffer.
+    pub fn wait(mut self) -> Result<(Uhci<'d, Blocking>, Buf::View), Error> {
+        self.uhci.uart.tx.flush()?;
+        while !self.is_done() {}
+        self.uhci.channel.tx.stop_transfer();
+        let retval = unsafe {
+            (
+                ManuallyDrop::take(&mut self.uhci),
+                ManuallyDrop::take(&mut self.dma_buf),
+            )
+        };
+        core::mem::forget(self);
+        Ok(retval)
+    }
+}
+
+impl<'d, Buf: DmaTxBuffer> UhciDmaTxTransfer<'d, Async, Buf> {
+    /// Waits for the DMA transfer to complete.
+    ///
+    /// This method blocks until the transfer is finished and returns the
+    /// `Uhci` instance and the associated buffer.
+    pub async fn wait(mut self) -> Result<(Uhci<'d, Async>, Buf::View), Error> {
+        // Workaround for an issue when it doesn't actually wait for the transfer to complete. I'm
+        // lost at this point, this is the only thing that worked
+        self.uhci.uart.tx.flush_async().await?;
+        DmaTxFuture::new(&mut self.uhci.channel.tx).await?;
+
+        self.uhci.channel.tx.stop_transfer();
+        let retval = unsafe {
+            (
+                ManuallyDrop::take(&mut self.uhci),
+                ManuallyDrop::take(&mut self.dma_buf),
+            )
+        };
+        core::mem::forget(self);
+        Ok(retval)
+    }
+}
+
+impl<Dm: DriverMode, Buf: DmaTxBuffer> Deref for UhciDmaTxTransfer<'_, Dm, Buf> {
+    type Target = Buf::View;
+
+    fn deref(&self) -> &Self::Target {
+        &self.dma_buf
+    }
+}
+
+impl<Dm: DriverMode, Buf: DmaTxBuffer> DerefMut for UhciDmaTxTransfer<'_, Dm, Buf> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.dma_buf
+    }
+}
+
+impl<Dm, Buf> Drop for UhciDmaTxTransfer<'_, Dm, Buf>
+where
+    Dm: DriverMode,
+    Buf: DmaTxBuffer,
+{
+    fn drop(&mut self) {
+        self.uhci.channel.tx.stop_transfer();
+
+        unsafe {
+            ManuallyDrop::drop(&mut self.uhci);
+            ManuallyDrop::drop(&mut self.dma_buf);
+        }
+    }
+}
+
+// Based on SpiDmaTransfer
+/// A structure representing a DMA transfer for UHCI/UART.
+///
+/// This structure holds references to the UHCI instance, DMA buffers, and
+/// transfer status.
+pub struct UhciDmaRxTransfer<'d, Dm, Buf>
+where
+    Dm: DriverMode,
+    Buf: DmaRxBuffer,
+{
+    uhci: ManuallyDrop<Uhci<'d, Dm>>,
+    dma_buf: ManuallyDrop<Buf::View>,
+}
+
+impl<'d, Buf: DmaRxBuffer, Dm: DriverMode> UhciDmaRxTransfer<'d, Dm, Buf> {
+    fn new(uhci: Uhci<'d, Dm>, dma_buf: Buf) -> Self {
+        Self {
+            uhci: ManuallyDrop::new(uhci),
+            dma_buf: ManuallyDrop::new(dma_buf.into_view()),
+        }
+    }
+
+    /// Returns true when [Self::wait] will not block.
+    pub fn is_done(&self) -> bool {
+        self.uhci.channel.rx.is_done()
+    }
+
+    /// Cancels the DMA transfer.
+    pub fn cancel(mut self) -> (Uhci<'d, Dm>, Buf) {
```

Comment:
You'll likely need to rebase or merge for this one to compile
```suggestion
    pub fn cancel(mut self) -> (Uhci<'d, Dm>, Buf::Final) {
```